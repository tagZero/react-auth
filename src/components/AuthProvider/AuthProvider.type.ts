export type NotificationType = 'info' | 'warn' | 'error' | 'success';

export interface LoginPropsType {
  email: string;
  password: string;
}

export interface RegisterPropsType extends LoginPropsType {
  firstName: string;
  lastName: string;
  birthDate: number;
}

export interface AuthProviderPropsType {
  isAuthenticated?: () => boolean;
  login?: (props: LoginPropsType, context?: AuthProviderType) => Promise<any>;
  logout?: (context?: AuthProviderType) => Promise<any>;
  register?: (props: RegisterPropsType, context?: AuthProviderType) => Promise<any>;
  resetPassword?: (props: { email: string }, context?: AuthProviderType) => Promise<any>;
  modules?: any;
  options?: any;
}

export interface AuthProviderType extends AuthProviderPropsType {
  authData?: any;
  getModulePath: (module: string) => string;
  notification?: any;
  setAuthData?: (props: any) => any;
  setModules?: (props: any) => any;
  setNotification?: (props: NotificationState) => void;
  setOptions?: (props: any) => any;
  setToken?: (token: string | null) => any;
  token?: string | null;
}

export interface NotificationState {
  type: NotificationType;
  message: string;
}
