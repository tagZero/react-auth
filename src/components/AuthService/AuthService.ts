import { AuthServiceType } from './AuthService.type';

class AuthServiceClass implements AuthServiceType {
  authData;
  modules;
  options;
  setNotification;
  setAuthData;
  history;

  constructor(props?: { options?: any; modules?: any }) {
    if (props?.options?.pathPrefix) {
      if (props.options.pathPrefix[0] !== '/') {
        throw new Error(
          'AuthService pathPrefix should be empty string "" or it should start with "/" character.'
        );
      }
    }

    this.options = {
      ...props?.options,
      pathPrefix: props?.options.pathPrefix || '/auth'
    };
    this.modules = {
      ...props?.modules,
      login: {
        path: '/login',
        title: 'Login',
        text: 'I have an account',
        description: 'Please provide your credentials below to login.',
        ...props?.modules?.login
      },
      register: {
        path: '/register',
        title: 'Register',
        text: 'Create new account',
        description: 'Please provide information below to register a new user.',
        passwordPattern: '(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[!@#$%&*?])[A-Za-z\\d!@#$%&*?]{8,}',
        passwordPatternMessage:
          'Please enter min. 8 characters having at least 1 capital letter, 1 lowercase letter, 1 number and 1 special character (!@#$%&*?)',
        ...props?.modules?.register
      },
      resetPassword: {
        path: '/reset-password',
        title: 'Reset password',
        text: 'Forgot password?',
        description: 'Please provide your e-mail address to reset your password.',
        ...props?.modules?.resetPassword
      }
    };
  }

  getModulePath = (module: string) => {
    if (this.options.pathPrefix === '/') {
      return this.modules[module].path;
    }
    return this.options.pathPrefix + this.modules[module].path;
  };

  resetPassword = async ({ email }: { email: string }) => {
    return Promise.resolve();
  };

  isAuthenticated = (): boolean => {
    return false;
  };

  login = async ({ email, password }: { email: string; password: string }) => {
    return Promise.resolve();
  };

  logout = (): Promise<any> => {
    return Promise.resolve();
  };

  register = async ({ firstName, lastName, email, password }: any) => {
    return Promise.resolve();
  };
}

export default AuthServiceClass;
