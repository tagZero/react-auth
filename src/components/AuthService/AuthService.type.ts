export interface AuthServiceType {
  authData: any;
  modules: any;
  options: any;
  isAuthenticated: () => boolean;
  resetPassword: (props: any) => Promise<any>;
  login: (props: any) => Promise<any>;
  register: (props: any) => Promise<any>;
  logout: () => Promise<any>;
  getModulePath: (module: string) => string;
}
