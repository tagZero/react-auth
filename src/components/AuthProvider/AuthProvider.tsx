import React, { createContext, useContext, useState } from 'react';
import { useHistory } from 'react-router-dom';
import { AuthServiceType, NotificationState } from '../AuthService/AuthService.type';

const AuthContext = createContext({} as AuthServiceType);
export const useAuth = () => useContext(AuthContext);

const AuthProvider: React.FC<{ service: AuthServiceType }> = ({ service, children }) => {
  const [authData, setAuthData] = useState();
  const [notification, setNotification] = useState<NotificationState>();
  const history = useHistory();

  const {
    login,
    logout,
    register,
    resetPassword,
    isAuthenticated,
    modules,
    options,
    getModulePath
  } = service;

  service.setNotification = setNotification;
  service.setAuthData = setAuthData;
  service.history = history;

  return (
    <AuthContext.Provider
      value={{
        authData,
        isAuthenticated,
        login,
        logout,
        register,
        resetPassword,
        modules,
        options,
        getModulePath,
        notification
      }}
    >
      {children}
    </AuthContext.Provider>
  );
};

export default AuthProvider;
