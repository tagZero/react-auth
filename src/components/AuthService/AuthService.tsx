import { AuthServiceType } from './AuthService.type';

class AuthServiceClass implements AuthServiceType {
  authData: any = null;
  modules: any = null;
  options: any = null;

  constructor(props?: { options?: any; modules?: any }) {
    if (props?.options?.pathPrefix) {
      if (props.options.pathPrefix[0] !== '/') {
        throw new Error(
          'AuthService pathPrefix should be empty string "" or it should start with "/" character.'
        );
      }
    }

    this.options = {
      ...props?.options,
      pathPrefix: props?.options.pathPrefix || '/auth'
    };
    this.modules = {
      ...props?.modules,
      login: {
        path: '/login',
        title: 'Login',
        text: 'I have an account',
        description: 'Please provide your credentials below to login.',
        ...props?.modules?.login
      },
      register: {
        path: '/register',
        title: 'Register',
        text: 'Create new account',
        description: 'Please provide information below to register a new user.',
        passwordPattern: '(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[!@#$%&*?])[A-Za-z\\d!@#$%&*?]{8,}',
        passwordPatternMessage:
          'Please enter min. 8 characters having at least 1 capital letter, 1 lowercase letter, 1 number and 1 special character (!@#$%&*?)',
        ...props?.modules?.register
      },
      resetPassword: {
        path: '/reset-password',
        title: 'Reset password',
        text: 'Forgot password?',
        description: 'Please provide your e-mail address to reset your password.',
        ...props?.modules?.resetPassword
      }
    };
  }

  getModulePath = (module: string) => {
    if (this.options.pathPrefix === '/') {
      return this.modules[module].path;
    }
    return this.options.pathPrefix + this.modules[module].path;
  };

  resetPassword = ({ email }: { email: string }): Promise<any> => {
    console.log({ email });
    return Promise.resolve();
  };

  isAuthenticated = (): boolean => {
    return false;
  };

  login = ({ email, password }: { email: string; password: string }): Promise<any> => {
    console.log({ email, password });
    return Promise.resolve();
  };

  logout = (): Promise<any> => {
    return Promise.resolve();
  };

  register = ({ firstName, lastName, email, password }: any): Promise<any> => {
    console.log({ firstName, lastName, email, password });
    return Promise.resolve();
  };
}

export default AuthServiceClass;
