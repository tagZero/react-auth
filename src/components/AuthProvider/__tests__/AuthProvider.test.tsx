import React from 'react';
import { render } from '@testing-library/react';
import AuthProvider from '../AuthProvider';
import AuthService from '../../AuthService/AuthService';

describe('AuthProvider', () => {
  const service = new AuthService();
  const tree = () =>
    render(
      <AuthProvider service={service}>
        <div>Test</div>
      </AuthProvider>
    );

  test('renders properly', () => {
    const { asFragment } = tree();
    expect(asFragment()).toMatchSnapshot();
  });
});
