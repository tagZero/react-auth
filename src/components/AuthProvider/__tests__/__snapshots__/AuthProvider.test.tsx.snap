// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthProvider renders properly 1`] = `
<DocumentFragment>
  <div>
    Test
  </div>
</DocumentFragment>
`;
